gcc error compiling generated code using `for` with iterator

### Description

```nim
iterator p(): int =
  when nimvm:
    yield 0
  else:
    yield 0

for k in p():
  let m = ""
  proc e() = discard m & ""
  e()
```

### Nim Version

```
Nim Compiler Version 1.6.10 [Linux: amd64]
Compiled at 2022-11-23
Copyright (c) 2006-2021 by Andreas Rumpf

active boot switches: -d:release
```

```
Nim Compiler Version 1.6.11 [Linux: amd64]
Compiled at 2022-12-14
Copyright (c) 2006-2021 by Andreas Rumpf

git hash: 7f90bcf5b4cbe5b7da534dd79bafbb4cafb313fa
active boot switches: -d:release
```

```
Nim Compiler Version 1.7.3 [Linux: amd64]
Compiled at 2022-12-15
Copyright (c) 2006-2022 by Andreas Rumpf

git hash: 8054be6e529db9e65a18e4c4a37ec54f64f9121c
active boot switches: -d:release
```

```
gcc (Debian 12.2.0-10) 12.2.0
Copyright (C) 2022 Free Software Foundation, Inc.
```

### Current Output

```text
/home/user/.cache/nim/os_d/@mos.nim.c: In function ‘e__os_4’:
/home/user/.cache/nim/os_d/@mos.nim.c:125:28: error: ‘m__os_3’ undeclared (first use in this function); did you mean ‘m__os_6’?
  125 |         T1_ = rawNewString(m__os_3.len + 0);
      |                            ^~~~~~~
      |                            m__os_6
/home/user/.cache/nim/os_d/@mos.nim.c:125:28: note: each undeclared identifier is reported only once for each function it appears in
Error: execution of an external compiler program 'gcc -c  -w -fmax-errors=3 -pthread   -I/home/user/Nim/lib -I/home/user -o /home/user/.cache/nim/os_d/@mos.nim.c.o /home/user/.cache/nim/os_d/@mos.nim.c' failed with exit code: 1
```


### Expected Output

```text
Not a gcc error
```


### Possible Solution

_No response_

### Additional Information

_No response_

--- a/compiler/transf.nim
+++ b/compiler/transf.nim
@@ -324,6 +324,14 @@ proc introduceNewLocalVars(c: PTransf, n: PNode): PNode =
     if a.kind == nkSym:
       n[1] = transformSymAux(c, a)
     return n
+  of nkProcDef: # todo optimize nosideeffects?
+    result = newTransNode(n)
+    let x = freshVar(c, n[namePos].sym)
+    idNodeTablePut(c.transCon.mapping, n[namePos].sym, x)
+    result[namePos] = x # we have to copy proc definitions for iters
+    for i in 1..<n.len:
+      result[i] = introduceNewLocalVars(c, n[i])
+    result[namePos].sym.ast = result
   else:
     result = newTransNode(n)
     for i in 0..<n.len:

--- a/tests/iter/titer.nim
+++ b/tests/iter/titer.nim
@@ -112,3 +112,18 @@ let res = collect:
       fn2(v2)
 
 doAssert res == @[42, 43, 43, 44]
+
+block: # bug #21110
+  iterator p(): int =
+    when nimvm:
+      yield 0
+    else:
+      yield 0
+
+  template foo =
+    for k in p():
+      let m = ""
+      proc e() = discard m & ""
+      e()
+  static: foo()
+  foo()

